### PR TITLE
feat: thirdparty rework

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `webJsInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: lib/ts/recipe/thirdparty/types.ts [hunk 1]
```diff
@@ -24,12 +24,14 @@ import {
...
 
-export type PreAndPostAPIHookAction = "GET_AUTHORISATION_URL" | "THIRD_PARTY_SIGN_IN_UP";
+export type PreAndPostAPIHookAction = "GET_AUTHORISATION_URL" | "THIRD_PARTY_SIGN_IN_UP" | "GET_PROVIDERS";
 
 export type PreAPIHookContext = RecipePreAPIHookContext<PreAndPostAPIHookAction>;
 export type PostAPIHookContext = RecipePostAPIHookContext<PreAndPostAPIHookAction>;
 
 export type UserInput = {
+    clientType?: string;
+    getTenantId?: () => Promise<string | undefined>;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/types.ts:34]: move getTenantId into the relevant functions -> getAuthorisationUrl, getProviders

File: lib/ts/recipe/thirdparty/types.ts [hunk 2]
```diff
@@ -70,6 +75,18 @@ export type ThirdPartyUserType = {
     };
 };
 
+export type ThirdPartyInput = {
+    thirdParty: {
+        clientType: string;
+        getTenantId: () => Promise<string | undefined>;
+    };
+};
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/types.ts:83]: no need for this type

File: lib/ts/recipe/thirdparty/types.ts [hunk 3]
```diff
@@ -70,6 +75,18 @@ export type ThirdPartyUserType = {
...
+
+export type ProviderInfo = {
+    id: string;
+    name?: string;
+};
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/types.ts:88]: don't create types specifically

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdparty/index.ts [hunk 4]
```diff
@@ -159,77 +99,15 @@ export default class RecipeWrapper {
...
-    /**
-     * Returns the auth state from the current URL
-     *
-     * @param userContext Refer to {@link https://supertokens.com/docs/thirdparty/advanced-customizations/user-context the documentation}
-     *
-     * @returns The "state" query param from the current URL. Returns an empty string if no state exists
-     */
-    static getAuthStateFromURL(input?: { userContext?: any }): string {
-        return Recipe.getInstanceOrThrow().recipeImplementation.getAuthStateFromURL({
+    static getProviders(input?: { tenantId?: string; userContext?: any }): Promise<{
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/index.ts:102]:
```suggestion
    static getConfiguredProviders(input?: { tenantId?: string; userContext?: any }): Promise<{
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdparty/utils.ts [hunk 5]
```diff
@@ -22,6 +22,7 @@ export function normaliseUserInput(config: InputType): NormalisedInputType {
     };
 
     return {
+        clientType: config.clientType || "web",
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/utils.ts:25]: Done

File: lib/ts/recipe/thirdparty/types.ts [hunk 6]
```diff
@@ -168,14 +180,25 @@ export type RecipeInterface = {
           }
     >;
 
+    getProviders: (input: { tenantId?: string; userContext?: any; options?: RecipeFunctionOptions }) => Promise<{
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/types.ts:183]: Done

File: lib/ts/recipe/thirdparty/types.ts [hunk 7]
```diff
@@ -44,6 +46,8 @@ export type UserInput = {
 export type InputType = AuthRecipeInputType<PreAndPostAPIHookAction> & UserInput;
 
 export type NormalisedInputType = AuthRecipeNormalisedInputType<PreAndPostAPIHookAction> & {
+    clientType: string;
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/types.ts:49]: Done

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 8]
```diff
@@ -14,15 +14,19 @@
...
 import STGeneralError from "../../error";
 import { PreAndPostAPIHookAction } from "./types";
 import { WindowHandlerReference } from "supertokens-website/utils/windowHandler";
 
 export default function getRecipeImplementation(
-    recipeImplInput: RecipeImplementationInput<PreAndPostAPIHookAction>
+    recipeImplInput: {
+        thirdParty: {
+            clientType: string;
+        };
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts:28]: Done

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 9]
```diff
@@ -240,12 +259,74 @@ export default function getRecipeImplementation(
...
+            if (input.tenantId !== undefined) params.tenantId = input.tenantId;
+
+            const { jsonBody, fetchResponse } = await querier.get<{
+                status: "OK";
+                providers: {
+                    id: string;
+                    name?: string;
+                }[];
+            }>(
+                "/tenant/providers",
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts:285]: Done

File: lib/ts/recipe/thirdparty/index.ts (hunk 4 above)
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/index.ts:102]: Done

File: lib/ts/recipe/thirdparty/types.ts (hunk 1 above)
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/types.ts]: Done

File: lib/ts/recipe/thirdparty/types.ts (hunk 2 above)
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/types.ts]: Done

File: lib/ts/recipe/thirdparty/types.ts (hunk 3 above)
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/types.ts]: Done

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 10]
```diff
@@ -240,12 +259,74 @@ export default function getRecipeImplementation(
...
+        }): Promise<{
+            status: "OK";
+            providers: {
+                id: string;
+                name?: string;
+            }[];
+            fetchResponse: Response;
+        }> {
+            const params: Record<string, string> = {};
+            if (input.tenantId !== undefined) params.tenantId = input.tenantId;
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts:276]: queryParams type doesn't allow this

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 11]
```diff
@@ -240,12 +259,74 @@ export default function getRecipeImplementation(
...
+            const { jsonBody, fetchResponse } = await querier.get<{
+                status: "OK";
+                providers: {
+                    id: string;
+                    name?: string;
+                }[];
+            }>(
+                "/tenant/providers",
+                {},
+                params,
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts:287]: queryParams type doesn't allow this

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 12]
```diff
@@ -65,69 +69,80 @@ export default function getRecipeImplementation(
...
             url: string;
+            pkceCodeVerifier?: string;
             fetchResponse: Response;
         }> {
+            const params: Record<string, string> = {
+                thirdPartyId: input.thirdPartyId,
+                redirectURIOnProviderDashboard: input.redirectURIOnProviderDashboard,
+            };
+            params.clientType = recipeImplInput.thirdParty.clientType;
+            if (input.tenantId !== undefined) params.tenantId = input.tenantId;
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts:136]: queryParams type doesn't allow this

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 13]
```diff
@@ -174,9 +190,8 @@ export default function getRecipeImplementation(
                 userContext: input.userContext,
             });
 
-            const code = this.getAuthCodeFromURL({
-                userContext: input.userContext,
-            });
+            const queryParams = getAllQueryParams();
+            const queryParamsObj: any = Object.fromEntries(queryParams);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdparty/recipeImplementation.ts]: Done

File: test/with-typescript/index.ts [hunk 14]
```diff
@@ -266,8 +266,8 @@ function getThirdPartyFunctions(original: ThirdPartyRecipeInterface): ThirdParty
         generateStateToSendToOAuthProvider: function (input) {
             return original.generateStateToSendToOAuthProvider(input);
         },
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply test/with-typescript/index.ts:268]: Will fix this after the implementation of all interfaces (tptp and tpless), while updating tests.

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**